Allow submitting only the fields to update in `edit_states_daily` endpoint

space-buzzer opened the pull request:
Move the check for differences between submitted and existing values to the data model class.

Enable submitting only the fields to update, without repeating existing values for `/api/v1/batches/edit_states_daily` endpoint. This will be more relevant for automated, or though spreadsheets for workflows such as backfills.
Add a test for this behaviour.

**There are 2 commits in this PR:**
1. *The 1st commit* in this PR is beefing up `editdiff.py` (that I really want to rename to ChangeSet). So that this class, with all the info about the changes generates the field diff and date-range diff, number of changed rows, etc (everything that goes into messages).
1. *The 2nd commit* is to actually add the logic to handle sending only fields for update and not all fields.

zachlipton commented:
I'm going to ask @julia326 to take a closer look at this since she knows`edit_states_daily` better than I do, but my only question is about how we handle the case where fields are being intentionally blanked out. 

Right now, the DataStore and multi-edit tool omit empty columns from the JSON they post to `edit_states_daily` entirely (the multi-edit tool does check to make sure all fields from `states/daily` get mapped or are part of an exception list, to make sure we don't accidentally blank out a column if it's omitted from the edit tool). So currently a missing column means it needs to be blanked out, and I think we still need a way to signal that because it happens sometimes. 

Am I understanding correctly that if a field is completely missing with this patch, it won't be blanked out anymore? I'm fine with changing the clients to deal with that (sending `field: null` in the JSON for all blank comments presumably, though that's going to produce a much larger/slower output for the DataStore) or another option, but want to make sure we preserve the option to blank out fields when editing.

space-buzzer commented:
If the field exists in the request, it should be part of the new batch, which whatever value it had in the request -- blank or not. If it's not part of the request, we should carry over the previous value.
If we want to blank something (from DE side: e.g., moving Total People to Total Specimens) we send 2 columns: the old one with an empty value and the new one with the correct value. 

Why will it be larger/slower for DataStore?

zachlipton commented:
Ah sorry I think I've got it. What you'd do here is:

1. Field populated in the `edit_states_daily` request (CoreData JSON)—gets edited (or left alone if unchanged)
1. Field entirely missing from the request—value gets preserved from whatever it was before
1. Field explicitly set to `null`—value is blanked out

We should probably add a unit test to codify that explicitly. 

Right now, the DataStore and multi-edit tool are using #2 to mean "blank out the value"—if you clear the cell in the sheet, it will be missing from the JSON we generate and so we set it to null in the DB. So we can do this and switch to #3, with all blank spreadsheet cells explicitly set to `null` in the JSON, which would preserve the same behavior. It just means a larger JSON blob, but I think that should be ok.

space-buzzer commented:
There was an existing test that I updated and another one that I added, so now all the options are checked:
- Sending fields to update
- Not sending fields, they preserve their value
- Sending fields with empty values, they're updated to empty value

julia326 commented on app/models/data.py:
```diff
@@ -242,10 +242,38 @@ def totalTestResults(self):
+        other = CoreData(**dict_other)
+
+        # special casing for aliasse (maybe move elsewhere?)
+        if 'lastUpdateIsoUtc' in dict_other:
```
what if `lastUpdateTime` is also in `dict_other`? Probably worth only overriding if it doesn't contain `lastUpdateTime`

julia326 replied:
that makes sense. In that case may as well combine into a single if condition, `if 'lastUpdateIsoUtc' in dict_other and not 'lastUpdateTime' in dict_other`

julia326 commented on app/models/data.py:
```diff
@@ -242,10 +242,38 @@ def totalTestResults(self):
+        Return: List[str], empty list if the dict results in no changes to current data
+        '''
+        fields = []
+        if not dict_other:
```
would you expect `dict_other` to ever be empty, or would it be an error?

julia326 commented on app/models/data.py:
```diff
@@ -242,10 +242,38 @@ def totalTestResults(self):
+        return fields
 
+    @staticmethod
+    def _cleanup_kwargs(kwargs):
```
nit: could you rename to `_cleanup_date_kwargs` so we know this function is relatively date-specific? Kind of hoping we don't do any other kwarg cleanup anyway

julia326 commented on app/models/data.py:
```diff
@@ -264,9 +292,32 @@ def __init__(self, **kwargs):
+
+    def __init__(self, **kwargs):
+        # strip any empty string fields from kwargs
+        kwargs = {k: v for k, v in kwargs.items() if v is not None and v != ""}
```
why strip the empty string fields?

julia326 replied:
:-O :-O I *think*... my original idea was to let DB null values be null values?

julia326 commented on app/models/data.py:
```diff
@@ -264,9 +292,32 @@ def __init__(self, **kwargs):
             kwargs['date'] = date.today()
+        return kwargs
+
+    def copy_with_updates(self, **kwargs):
```
sorry if this is a silly question - why not just make a new `kwargs` dictionary and pass that to the `CoreData` constructor, instead of setting these values directly in this method? What would that break?

space-buzzer replied:
err, this becomes annoying because of SQLAlchemy and its special classes. I can look into it later, I haven't used SQLAlchemy before, but to do this type of copy and preserving the constructor I could expunge object to have a "clean version" not connected to a session that later I could add again as a new row

julia326 replied:
Since we have `to_dict()` methods on these objects, would it work to extract that, do updates within that dict and then make a new `CoreData` object from that updated dict as kwargs? Sorry if sending you on a wild goose chase, let me know if this gets annoying. Hoping to avoid a lot of SQLAlchemy magic in the API code in general. Is what you're trying to do ultimately to create a new CoreData object identical in field values to this one, with the exception of the overriding kwargs as passed in here?

space-buzzer replied:
Yeah, this seems to work. Donno whether I tried it and something didnt work in the past... can't recall

julia326 commented on tests/app/edit_test.py:
```diff
@@ -206,3 +206,56 @@ def test_edit_core_data_from_states_daily_timestamps_only(app, headers, slack_mo
+    assert "state: NY" in slack_mock.chat_postMessage.call_args[1]['text']
+    batch_id = resp.json['batch']['batchId']
+    assert resp.json['batch']['user'] == 'testing'
+    # we've changed only lastUpdateIsoUtc, which is lastUpdateTime on output
```
this comment is wrong here maybe

julia326 commented on tests/app/common.py:
```diff
@@ -48,13 +48,25 @@
     "positive": 16,
-    "negative": 4
+    "negative": 4,
+    "inIcuCurrently": None
```
which test(s) does blanking out `inIcuCurrently` affect and are they affected the way we'd expect?

julia326 reviewed (commented):
Thank you @space-buzzer ! This change mostly makes sense to me just a few clarifying questions around the specifics

julia326 commented on app/api/data.py:
```diff
@@ -375,63 +371,47 @@ def edit_core_data_from_states_daily():
+                changed_rows.append(changed_for_date)
+
+        # if any value in the row is different, make an edit batch
+        if is_edited or is_new:
```
small nit: may be slightly easier to read as

```
if is_edited:
  edited_core_data = data_for_date.copy_with_updates(**core_data_dict)
elif is_new:
  edited_core_data = CoreData(**core_data_dict)
  new_rows.append(edited_core_data)
```

julia326 reviewed (approved):
very close with this! Just a couple small things. I'm mildly nervous about that SQLAlchemy question, and would be good to explore that for a sec but beyond that this looks great and feel free to merge
